### ISSUE TITLE
Fusion map for upstream PR #878: Kimi-K3 MoE low-token decode

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#49]
# PR #878: Kimi-K3 MoE fusion map

Source: [lightseekorg/tokenspeed PR #878](https://github.com/lightseekorg/tokenspeed/pull/878), `perf: Optimize Kimi K3 MoE input projections and low-token decode`.

Short version: the PR reduces Kimi-K3's low-token MoE path by combining three input projections, co-dispatching the routed and shared expert outputs, and folding the two final additions into the routed up-projection. It also adds/tunes several supporting small-batch kernels.

## Fusion summary

| Original operations | New operation/kernel | Main implementation |
|---|---|---|
| Router projection + routed-down projection + shared gate/up projection + shared SiTU | `latent_moe_input_projections`; `gluon_latent_input_decode_gfx950` for M=1 | `tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/fp16/latent_input_decode.py` |
| The same three projections for M=2–128 | One packed GEMM weight pass + fused split epilogue via `gluon_latent_input_small_batch_gfx950` | `tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/fp16/latent_input_small_batch.py` |
| Routed W2 + route-weight combine, alongside a separate shared-down GEMM | One stage-2 launch containing routed-W2 and shared-down programs | `tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/situ_decode.py` |
| Routed up-projection, then `prefix + routed + shared` | Projection with both additions in its store epilogue | `tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/gemm/fp16/mm.py` and `tokenspeed-kernel/.../ops/gemm/triton_gemv.py` |

## 1. Packed MoE input projections

All three branches read the same `[M, 7168]` hidden state:

```text
                         -> router logits [M, 896]
hidden state [M, 7168]  -> routed latent [M, 3584]
                         -> shared gate/up [M, 1536] -> SiTU [M, 768]
```

### Original form

```python
router_logits = self.gate(hidden_states)
routed_input = self.routed_expert_down_proj(hidden_states)[0]

shared_raw = self.shared_experts.gate_up_proj(hidden_states)[0]
shared_gate, shared_up = shared_raw.chunk(2, dim=-1)
shared_input = situ_and_mul(shared_gate, shared_up)
```

These were three independent projections, followed by the shared SiTU activation.

### New form

```python
router_logits, routed_input, shared_input = latent_moe_input_projections(
    hidden_states,
    self.gate.weight,
    self.routed_expert_down_proj.weight,
    self.shared_experts.gate_up_proj.weight,
    gate_clamp=self.shared_experts.act_fn.beta,
    up_clamp=self.shared_experts.act_fn.linear_beta,
)
```

At model load, the three weights are backed by consecutive rows of one allocation:

```python
self.packed_input_projection_weight = torch.cat(
    [router_weight, routed_down_weight, shared_gate_up_weight],
    dim=0,
)
```

This allows one packed GEMM to stream the combined weight rather than launching three GEMMs. The outputs retain their required types: FP32 router logits and BF16 routed/shared inputs.

### Selected kernels

For exactly one token, one Gluon launch performs all three projections and the shared SiTU epilogue:

```python
gluon_latent_input_decode_gfx950(
    hidden_states,
    router_weight,
    routed_down_weight,
    shared_gate_up_weight,
    beta=gate_clamp,
    linear_beta=up_clamp,
)
```

For M=2–128, the selected Gluon path uses:

```text
_packed_projection_gemm_kernel  # one packed weight pass
_split_epilogue_kernel          # FP32 router output, BF16 routed output, shared SiTU
```

This is one semantic dispatch but two physical kernel launches because split-K partials and the mixed-output epilogue are separate.

For larger supported token counts, `triton_latent_input_packed` similarly uses a packed projection kernel followed by a SiTU kernel.

- Public API and Torch fallback: `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/latent_input.py`
- Gluon registrations: `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/latent_input.py`
- Portable Triton path: `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/triton/latent_input.py`
- Runtime packing/call site: `python/tokenspeed/runtime/models/kimi_k3.py`
- Runtime orchestration: `python/tokenspeed/runtime/layers/moe/latent.py`

## 2. Routed W2 and shared-down co-dispatch

The routed expert computation already had two fused stages before this PR:

```text
stage 1: routed W13 projection + SiTU
stage 2: routed W2 projection + top-k route weighting + local-route combine
```

PR #878 adds the independent shared-expert down projection to the stage-2 launch.

### Original form

```python
# Existing routed expert path.
routed_intermediate = routed_w13_and_situ(routed_input, topk_ids)
routed_output = routed_w2_weight_and_combine(
    routed_intermediate,
    topk_ids,
    topk_weights,
)

# Separate BF16 shared-expert GEMM.
shared_output = kimi3_shared_down_projection(
    shared_input,
    self.shared_experts.down_proj.weight,
)
```

### New form

```python
routed_output, shared_output = latent_moe_expert_shared(
    routed_input,
    w13_weight,
    w13_scale,
    w2_weight,
    w2_scale,
    topk_weights,
    topk_ids,
    shared_input,
    shared_down_weight,
    routed_out=routed_output,
    shared_out=shared_output,
    ...,
)
```

The selected implementation is:

```text
gluon_latent_expert_shared_gfx950
  -> gluon_latent_expert_shared_decode_gfx950
  -> gluon_a16w4_situ_warp_decode_ep_gfx950
  -> _stage2_a16w4_warp_gemv_combine
```

The expanded stage-2 grid contains two kinds of programs:

```python
if pid < NUM_ROUTED_PROGRAMS:
    # MXFP4 routed W2 + route-weight combine
else:
    # BF16 shared-down projection
```

This is launch fusion/co-dispatch, not arithmetic fusion between the branches: routed and shared results remain separate tensors so `all_reduce_two` can jointly reduce them afterward.

- Public orchestration API: `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/latent_decode.py`
- Registration: `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/latent_decode.py`
- Joint wrapper: `tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/latent_shared_decode.py`
- Actual fused stage-2 grid: `tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/situ_decode.py`
- Runtime call site: `python/tokenspeed/runtime/models/kimi_k3.py` (`_forward_fused_decode_pipeline`)

## 3. Final routed projection + two additions

After the routed/shared collective, Kimi-K3 must compute:

```text
prefix_sum + routed_up(routed_latent) + shared_output
```

### Original form

```python
routed_output = self.routed_expert_up_proj(routed_latent)[0]
output = add3(prefix_sum, routed_output, shared_output)
```

The projection materialized `routed_output`, then a separate `add3` kernel read all three tensors.

### New form

```python
output = self.routed_expert_up_proj.forward_add3(
    routed_latent,
    prefix_sum,
    shared_output,
)
```

which calls:

```python
kimi3_latent_projection_add3(
    routed_latent,
    routed_up_weight,
    prefix_sum,
    shared_output,
)
```

For M=1, `rowcta_gemv_add3` adds the two full-width tensors in the GEMV store epilogue. For M=16 on `gfx950`, this PR adds:

```python
gluon_mm_a16w16_add3_m16_gfx950(
    routed_latent,
    routed_up_weight,
    prefix_sum,
    shared_output,
)
```

Conceptually, the fused store is:

```python
output = matmul_accumulator + prefix_sum + shared_output
```

Other shapes use the safe composed projection plus `add3` fallback.

- Dispatch and fallback: `tokenspeed-kernel/python/tokenspeed_kernel/ops/gemm/kimi3.py` (`kimi3_latent_projection_add3`)
- M=16 Gluon implementation: `tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/gemm/fp16/mm.py`
- M=1 row-CTA implementation: `tokenspeed-kernel/python/tokenspeed_kernel/ops/gemm/triton_gemv.py`
- Runtime wrapper: `python/tokenspeed/runtime/layers/moe/latent.py` (`Kimi3LatentProjection.forward_add3`)

## 4. General one-token routing fusion added by the PR

The PR also adds a one-CTA routing kernel which combines:

```text
sigmoid(logits)
  + correction bias for selection
  + top-k selection
  + gather unbiased sigmoid weights
  + optional weight normalization
  + routed scaling
  + optional logical-to-physical expert mapping
```

Original/reference form:

```python
scores = router_logits.sigmoid()
topk_ids = torch.topk(scores + correction_bias, topk).indices
topk_weights = scores.gather(1, topk_ids)
topk_weights /= topk_weights.sum(dim=-1, keepdim=True)
topk_weights *= routed_scaling_factor
topk_ids = logical_to_physical_map[topk_ids]
```

New fused form:

```python
topk_weights, topk_ids = moe_sigmoid_bias_topk(
    router_logits,
    correction_bias,
    topk,
    routed_scaling_factor=routed_scaling_factor,
    logical_to_physical_map=logical_to_physical_map,
)
```

Registrations:

```text
triton_decode_sigmoid_bias_topk
triton_decode_sigmoid_bias_topk_mapped
```

- Implementation: `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/triton/decode_sigmoid_topk.py`
- Selection API: `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/sigmoid_topk.py`

Important nuance: the exact Kimi-K3 M=1 shape (`[1, 896]`, top-k 16) already had the specialized `kimi3_sigmoid_bias_topk` packed-key kernel and continues to hit that earlier fast path. The new registrations generalize fused one-token routing to other eligible AMD shapes; they are not an additional launch removed from the exact K3 path.

### Before

```python
router_logits = router(hidden_states)
routed_input = routed_down(hidden_states)

shared_input = situ(shared_gate_up(hidden_states))
shared_output = shared_down(shared_input)

topk = sigmoid_bias_topk(router_logits)
routed_latent = routed_w13_situ_w2_combine(routed_input, topk)

shared_output, routed_latent = all_reduce_two(shared_output, routed_latent)
routed_output = routed_up(routed_latent)
output = add3(prefix_sum, routed_output, shared_output)
```

### New

```python
router_logits, routed_input, shared_input = latent_moe_input_projections(...)
topk = sigmoid_bias_topk(router_logits)

routed_latent, shared_output = latent_moe_expert_shared(
    routed_input,
    ...,
    shared_input,
    shared_down_weight,
)

shared_output, routed_latent = all_reduce_two(shared_output, routed_latent)
output = kimi3_latent_projection_add3(
    routed_latent,
    routed_up_weight,
    prefix_sum,
    shared_output,
)
```

## Optimizations in the PR that are not new fusions

- The routed W13+SiTU stage was already fused. PR #878 widens/tunes its K tile and masks the K tail.
- The routed W2+top-k weighting+local combine stage was already fused. The PR tunes the M=4 schedule and adds shared-down work to its launch for M=1.
- `all_reduce_two` was already the joint routed/shared collective mechanism; the PR arranges the fused decode outputs so it can still be used.
- Packing the three input weights is an enabling layout change, not a mathematical fusion by itself.

The largest structural win is the input side: three large projections become one packed weight pass. The M=1 path additionally saves launches by co-dispatching shared-down with routed W2 and folding the final two additions into the routed-up projection.

**Comment by raikonenfnu** [comment]
## Short PR message

### Summary

Optimize Kimi-K3 MoE input projection and low-token decode on `gfx950` by reducing redundant weight passes, co-dispatching independent expert work, and folding final additions into the routed up-projection.

### Overall pipeline

```text
Before:
3 input projections -> top-k -> routed experts + separate shared expert
                    -> joint reduce -> routed up-projection -> add3

After:
packed input projection -> top-k -> routed W2/shared-down co-dispatch
                        -> joint reduce -> routed up-projection + add3 epilogue
```

### Fusion 1: MoE input projections

```text
Before:
hidden -> router GEMM --------------------------> FP32 router logits
hidden -> routed-down GEMM --------------------> BF16 routed latent
hidden -> shared gate/up GEMM -> shared SiTU ---> BF16 shared input

After:
hidden -> packed input-projection dispatch ----> router logits
                                           +---> routed latent
                                           +---> activated shared input
```

- M=1: `gluon_latent_input_decode_gfx950` performs the three projections and shared SiTU in one launch.
- M=2–128: `gluon_latent_input_small_batch_gfx950` uses one packed MFMA weight pass plus a mixed-output epilogue.
- Larger supported batches use the portable packed Triton path.

### Fusion 2: one-token routing operations

```text
Before:
sigmoid(logits) -> add correction bias -> top-k ids
                -> gather unbiased weights -> normalize -> scale
                -> optional logical-to-physical expert mapping

After:
logits + bias + optional map -> one packed-key top-k dispatch
                            -> final weights + physical expert ids
```

The PR adds `triton_decode_sigmoid_bias_topk` and its mapped variant. The exact Kimi-K3 M=1 `[1, 896]`, top-k 16 case already uses the earlier `kimi3_sigmoid_bias_topk` fast path; this new kernel generalizes the fusion to other eligible one-token AMD shapes.

### Fusion 3: routed W2 and shared-down co-dispatch

```text
Before:
routed W13 + SiTU -> routed W2 + route weighting + local combine ---> routed output
shared gate/up + SiTU -> separate shared-down GEMM ----------------> shared output

After:
routed W13 + SiTU -> one stage-2 launch:
                       routed W2 + route weighting + local combine -> routed output
                       shared-down projection ---------------------> shared output
```

`gluon_latent_expert_shared_gfx950` adds shared-down programs to the existing routed stage-2 grid. This is launch fusion/co-dispatch rather than arithmetic fusion between the branches; the two outputs remain separate for `all_reduce_two`.

### Fusion 4: routed up-projection and final additions

```text
Before:
routed latent -> routed up-projection -> materialized routed output
prefix + routed output + shared output -> separate add3 kernel -> output

After:
routed latent -> routed up-projection
                 + prefix + shared output in the store epilogue -> output
```

`kimi3_latent_projection_add3` selects `rowcta_gemv_add3` for M=1 and the new `gluon_mm_a16w16_add3_m16_gfx950` for M=16. Other shapes keep the composed fallback.

### Original form

```python
router_logits = self.gate(hidden_states)
routed_input = self.routed_expert_down_proj(hidden_states)[0]

shared_input = situ(self.shared_experts.gate_up_proj(hidden_states)[0])
shared_output = self.shared_experts.down_proj(shared_input)[0]

topk_output = self.topk(hidden_states, router_logits)
routed_latent = self.experts(routed_input, topk_output=topk_output, ...)
shared_output, routed_latent = all_reduce_two(shared_output, routed_latent)

routed_output = self.routed_expert_up_proj(routed_latent)[0]
output = add3(prefix_sum, routed_output, shared_output)
```

### New form

```python
router_logits, routed_input, shared_input = latent_moe_input_projections(
    hidden_states,
    self.gate.weight,
    self.routed_expert_down_proj.weight,
    self.shared_experts.gate_up_proj.weight,
    gate_clamp=gate_clamp,
    up_clamp=up_clamp,
)
topk_output = self.topk(hidden_states, router_logits)

routed_latent, shared_output = latent_moe_expert_shared(
    routed_input,
    w13_weight,
    w13_scale,
    w2_weight,
    w2_scale,
    topk_output.topk_weights,
    topk_output.topk_ids,
    shared_input,
    shared_down_weight,
    ...,
)
shared_output, routed_latent = all_reduce_two(shared_output, routed_latent)

output = kimi3_latent_projection_add3(
    routed_latent,
    routed_up_weight,
    prefix_sum,
    shared_output,
)
```

Weights are packed once at load time while each module retains a contiguous row view, so unsupported layouts and shapes can still use the existing composed fallbacks.

### Main files

- `tokenspeed-kernel-amd/.../moe/fp16/latent_input_{decode,small_batch}.py`
- `tokenspeed-kernel-amd/.../moe/mxfp4/{latent_shared_decode,situ_decode}.py`
- `tokenspeed-kernel-amd/.../gemm/fp16/mm.py`
- `tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/{latent_input,latent_decode}.py`
- `python/tokenspeed/runtime/{layers/moe/latent.py,models/kimi_k3.py}`

### Result

Reported 4K/1K throughput improvements on 8x MI350X, TP8/EP8:

- M=1: **+11.03%**
- M=2: **+15.01%**
- M=4: **+14.06%**
- M=8: **+7.27%**
- M=16: **+2.62%**

Upstream PR: https://github.com/lightseekorg/tokenspeed/pull/878
